src/lib/LocalMedia.ts

```diff
@@ -0,0 +1,183 @@
+    microphoneDevices: MediaDeviceInfo[];
+    speakerDevices: MediaDeviceInfo[];
+};
+type StreamUpdatedEvent = {
+    stream: MediaStream;
+};
+
+interface LocalMediaEventsMap {
+    camera_enabled: CustomEvent<CameraEnabledEvent>;
+    device_list_updated: CustomEvent<DeviceListUpdatedEvent>;
+    microphone_enabled: CustomEvent<MicrophoneEnabledEvent>;
+    stream_updated: CustomEvent<StreamUpdatedEvent>;
+}
+
+interface LocalMediaEventTarget extends EventTarget {
+    addEventListener<K extends keyof LocalMediaEventsMap>(
+        type: K,
+        listener: (ev: LocalMediaEventsMap[K]) => void,
+        options?: boolean | AddEventListenerOptions
+    ): void;
+    addEventListener(
+        type: string,
+        callback: EventListenerOrEventListenerObject | null,
+        options?: EventListenerOptions | boolean
+    ): void;
+}
+
+const TypedLocalMediaEventTarget = EventTarget as { new (): LocalMediaEventTarget };
+
+export default class LocalMedia extends TypedLocalMediaEventTarget {
+    private _constraints: MediaStreamConstraints;
+    private _rtcManagers: RtcManager[];
+    public stream: MediaStream;
+
+    constructor(constraints: MediaStreamConstraints) {
+        super();
+        this._constraints = constraints;
+        this.stream = new MediaStream();
+        this._rtcManagers = [];
+
+        navigator.mediaDevices.addEventListener("devicechange", this._updateDeviceList.bind(this));
+    }
+
+    addRtcManager(rtcManager: RtcManager) {
+        this._rtcManagers.push(rtcManager);
+    }
+
+    removeRtcManager(rtcManager: RtcManager) {
+        this._rtcManagers = this._rtcManagers.filter((r) => r === rtcManager);
+    }
+
+    getCameraDeviceId() {
+        return this.stream.getVideoTracks()[0]?.getSettings().deviceId;
+    }
+
+    getMicrophoneDeviceId() {
+        return this.stream.getAudioTracks()[0]?.getSettings().deviceId;
+    }
+
+    isCameraEnabled() {
+        return !!this.stream.getVideoTracks()[0]?.enabled;
+    }
+
+    isMicrophoneEnabled() {
+        return !!this.stream.getAudioTracks()[0]?.enabled;
+    }
+
+    toggleCameraEnabled(enabled?: boolean) {
+        const videoTrack = this.stream.getVideoTracks()[0];
+        if (!videoTrack) {
+            return;
+        }
+
+        const newValue = enabled ?? !videoTrack.enabled;
+        videoTrack.enabled = newValue;
+
+        this.dispatchEvent(new CustomEvent("camera_enabled", { detail: { enabled: newValue } }));
+    }
+
+    toggleMichrophoneEnabled(enabled?: boolean) {
+        const audioTrack = this.stream.getAudioTracks()[0];
+        if (!audioTrack) {
+            return;
+        }
+
+        const newValue = enabled ?? !audioTrack.enabled;
+        audioTrack.enabled = newValue;
+
+        this.dispatchEvent(new CustomEvent("microphone_enabled", { detail: { enabled: newValue } }));
+    }
+
+    async setCameraDevice(deviceId: string) {
+        const newStream = await navigator.mediaDevices.getUserMedia({ video: { deviceId } });
```

Comment:
This could throw, is that desired behavior and should be caught by upper layer?

Reply:
This should be handled here I think, and we should rather emit an error event which the hook could use to update a `state` property to something error-ish.

Reply:
Added handling